src/Messenger/Transport/DsnParser.php

```diff
@@ -4,28 +4,131 @@
-            throw new \InvalidArgumentException($message, 1643988474);
+            throw new InvalidArgumentException($message, 1643988474);
+        }
+
+        $query = [];
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $query);
+        }
+
+        // check for extra keys in options
+        $optionsExtraKeys = array_diff(array_keys($options), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < count($optionsExtraKeys)) {
+            throw new InvalidArgumentException(sprintf(
+                'Unknown option found: [%s]. Allowed options are [%s].',
+                implode(', ', $optionsExtraKeys),
+                implode(', ', array_keys(self::DEFAULT_OPTIONS))
+            ));
+        }
+
+        // check for extra keys in query
+        $queryExtraKeys = array_diff(array_keys($query), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < count($queryExtraKeys)) {
+            throw new InvalidArgumentException(sprintf(
+                'Unknown option found in DSN: [%s]. Allowed options are [%s].',
+                implode(', ', $queryExtraKeys),
+                implode(', ', array_keys(self::DEFAULT_OPTIONS))
+            ));
+        }
+
+        $options = $query + $options + self::DEFAULT_OPTIONS;
+        /** @var array{
+         *      shared_access_key_name: string|null,
+         *      shared_access_key: string|null,
+         *      namespace: string|null,
+         *      entity_path: string|null,
+         *      subscription: string|null,
+         *      token_expiry: int|string,
+         *      receive_mode: AzureTransport::RECEIVE_MODE_*,
+         * } $options
+         */
+
+        $options['shared_access_key_name'] = $this->pickOption('shared_access_key_name', 'user', $parsedUrl, $options);
+        $options['shared_access_key'] = $this->pickOption('shared_access_key', 'pass', $parsedUrl, $options);
+        $options['namespace'] = $this->pickOption('namespace', 'host', $parsedUrl, $options);
+        $options['token_expiry'] = (int) $options['token_expiry'];
+
+        // Missing topic or queue name
+        if (null === $options['entity_path']) {
+            throw new InvalidArgumentException(
+                sprintf('Missing entity_path (queue or topic) for the "%s" transport.', $transportName),
+                1643989596
+            );
+        }
+
+        // Invalid receive mode
+        if (false === in_array($options['receive_mode'], self::RECEIVE_MODES, true)) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'Invalid "%s" receive_mode for the "%s" transport. It must be one of: %s.',
+                    $options['receive_mode'],
+                    $transportName,
+                    implode(', ', self::RECEIVE_MODES)
+                ),
+                1643994036
+            );
+        }
+
+        // @phpstan-ignore-next-line after all those array merges PHPstan gets confused about the result shape
```

Comment:
I don't really like cheating with PHPStan as it may backfire some day. I checked out your branch locally and the issue comes from the ` pickOption()` method returning non-string values + your PHPDoc comment specifying null values.
I'd return this instead:
```php
return [
    'shared_access_key_name' =>  $this->pickOption('shared_access_key_name', 'user', $parsedUrl, $options, $transportName),
    'shared_access_key' =>  $this->pickOption('shared_access_key', 'pass', $parsedUrl, $options, $transportName),
    'namespace' =>  $this->pickOption('namespace', 'host', $parsedUrl, $options, $transportName),
    'entity_path' =>  $options['entity_path'],
    'subscription' =>  $options['subscription'],
    'token_expiry' =>  (int) $options['token_expiry'],
    'receive_mode' =>  $options['receive_mode'],
];
```
Then you could delete lines 75-89. Check my other comment about `pickOption()` to understand why I added `$transportName` in the argument list.